### ISSUE TITLE
`func-names` error on "inferrable names"

### ISSUE THREAD

**Issue by ValYouW** [eslint/eslint#7739]
Hi,
Excuse me for bringing issue #7235 up again, but why Foo.prototype.bar = function() {}; is incorrect when using the `as-needed` option? while I understand that the function's name property would be empty, in callstacks this function will have the name Foo.bar, and I think this goes along with the TSC decision of "see the rule as ensuring names are present for debugging".
Also this is the behavior of JSCS `requireNamedUnassignedFunctions` (which is actually what I'm trying to restore)...
Thx

--- EDIT ---
I propose a rule change for `func-names` to have option to error only for unassigned functions, I am not sure whether this should be a change or a whole new rule.

**What rule do you want to change?**
`func-names`

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option

**Please provide some example code that this change will affect:**
Currently `func-names` require all function expressions to have names, even if the name can be inferred from the context. The new option will enforce the rule only if the function is unassigned, like callbacks:
```js
setTimeout(function() {
    throw new Error('I am truly anonymous');
}, 1000);
```
The code below is currently "incorrect" using the `func-names` rule, adding an option (say: `unassigned-only`) will make it "correct"
```js
function Foo() {}
Foo.prototype.baz = function() {
    throw new Error('I am not so anonymous, at least V8 knows who I am');
};
```

**What does the rule currently do for this code?**
Errors - saying a name is required

**What will the rule do after it's changed?**
Will not error.

As I said, maybe a new rule is more appropriate here (I am currently using a custom rule I wrote to achieve that, can use it as a base for a new rule).

**Comment by ljharb** [comment]
(It would have the inferred name of `bar`, not `Foo.bar`)

**Comment by ValYouW** [comment]
@ljharb This is what I get:
```js
> function Foo() {}
undefined
> Foo.prototype.bar = function() {throw new Error('ERR');}
[Function]
> var foo = new Foo();
undefined
> foo.bar();
Error: ERR
    at Foo.bar (repl:1:39)
    at repl:1:5
    at sigintHandlersWrap (vm.js:22:35)
    at sigintHandlersWrap (vm.js:96:12)
    at ContextifyScript.Script.runInThisContext (vm.js:21:12)
    at REPLServer.defaultEval (repl.js:313:29)
    at bound (domain.js:280:14)
    at REPLServer.runBound [as eval] (domain.js:293:12)
    at REPLServer.<anonymous> (repl.js:513:10)
    at emitOne (events.js:101:20)
```

**Comment by ljharb** [comment]
That's in the stack trace, which isn't currently governed by a spec - the name would be `foo.bar.name`.

**Comment by loganfsmyth** [comment]
@ljharb That form doesn't get an inferred names in the ES6 spec AFAIK.

**Comment by not-an-aardvark** [comment]
```js
class Foo{}

Foo.prototype.bar = function() {}

console.log(Foo.prototype.bar.name);

// output: (empty string)
```

**Comment by ljharb** [comment]
Right, I'm just saying that the name it *would* get would be `bar`, not `Foo.bar` - I recognize that it doesn't actually get an inferred name.

**Comment by ValYouW** [comment]
I'm sorry, I probably misused the term "inferred", what I meant is that while the function doesn't "officially" has a name, it still can be inferred using the parent element, at least that's what I remember from esprima.
Now the questions is whether that's enough to satisfy the rule goal as "ensuring names are present for debugging".

**Comment by ljharb** [comment]
@ValYouW "name inference" refers to the requirement in the ES6 language spec that in some cases, the function's "name" property will be set to the name of the variable or property name that they're assigned to.

The rule is only satisfied if the function's "name" property contains the name, because native debuggers aren't the only tools that do debugging - imagine, for example, code that reports code coverage information, and wants to include the function's name in the output.

**Comment by ValYouW** [comment]
@ljharb I believe that such tools would have to infer the name as well.
Personally I don't see a difference between:
`var f = function() {}`
and
`Foo.prototype.bar = function() {}`
IMHO the `name` property is irrelevant when coming to determine whether a function is anonymous or not, but that's just me...

Another example, what the difference between:
```js
> var foo;
> foo = {boo: function(){}}
> foo.boo.name
'boo'
>
> var zoo = {};
> zoo.boo = function() {}
> zoo.boo.name
''
```

I think that when it comes to linting, both scenarios are the same, no?

**Comment by not-an-aardvark** [comment]
> I think that when it comes to linting, both scenarios are the same, no?

No, because the whole point of the `as-needed` option is to check whether the function would have an inferred `name` property. Since the function will have a `name` property in one of the examples but not the other, it's correct for the rule to only report one of those cases but not the other.

**Comment by ValYouW** [comment]
Agrre. Can there be an `anonymous-only` option that will error only when the function is truly anonymous? Practically speaking, what debugging tools won't show a name for `Foo.prototype.bar`?

**Comment by ljharb** [comment]
There's no spec governing debugging tools, so there's literally no way to know when something is inferred. Note that it's not just node/v8, but you'd need to check every browser, as well as JS-only debugging tools (which can't inspect the original source code, and thus, couldn't possible infer anything besides the `.name` property)

**Comment by ValYouW** [comment]
Do you think this should be a new rule/part of jscs compatability/discarded?

**Comment by ljharb** [comment]
My opinion (as a non-TSC member) is that it should be simply discarded. However, I believe the eslint team has committed to full jscs compatibility, so it seems appropriate to address it somehow.

**Comment by nzakas** [comment]
@ValYouW if you'd like to propose a rule change, please update the deescriptiob of this issue with a proposal using this template: https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md

**Comment by ValYouW** [comment]
@nzakas Done, not sure whether a change to rule is needed or a new rule (or discarded altogether, I am using a custom rule for now)

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.

**Comment by ValYouW** [comment]
In case anyone is interested in JSCS `requireNamedUnassignedFunctions` behavior I created an ESLint plugin:
https://github.com/ValYouW/eslint-plugin-named-unassigned-functions
